plugins/catalog-backend/src/processing/util.ts

```diff
@@ -86,3 +86,50 @@ export function isObject(value: JsonValue | undefined): value is JsonObject {
 export const validateEntity = entitySchemaValidator();
 
 export const validateEntityEnvelope = entityEnvelopeSchemaValidator();
+
+/**
+ * Creates a barrier with a timeout, that can be awaited or prematurely
+ * released either manually or by an abort signal.
+ */
+export function createBarrier(options: {
+  waitTimeoutMillis: number;
+  signal: AbortSignal;
+}): {
+  wait: () => Promise<void>;
+  release: () => void;
+} {
+  const { waitTimeoutMillis, signal } = options;
+  const resolvers: Array<() => void> = [];
+
+  function wait() {
+    if (signal.aborted || !(waitTimeoutMillis > 0)) {
+      return Promise.resolve();
+    }
+
+    return new Promise<void>(resolve => {
+      const timeoutHandle = setTimeout(done, waitTimeoutMillis);
+
+      function done() {
+        const index = resolvers.indexOf(done);
+        if (index !== -1) {
+          resolvers.splice(index, 1);
+        }
+        clearTimeout(timeoutHandle);
+        resolve();
+      }
+
+      resolvers.push(done);
+    });
+  }
+
+  function release() {
+    const resolversToCall = resolvers.splice(0, resolvers.length);
+    for (const resolver of resolversToCall) {
+      resolver();
+    }
+  }
+
+  signal.addEventListener('abort', release);
```

Comment:
Probably safest to remove this afterwards

Reply:
There is no real afterwards though.

Reply:
Moved this into a dedicated destroy function that gets called when the pipeline is stopped, for completeness